File: sdk/js/src/core/wire.ts
/**
 * GRAFT Wire Format — Packet encode/decode
 * Matches firmware/src/graft/core/graft_wire.{h,c}
 *
 * Packet layout:
 *   MAGIC(2) + VER(1) + TYPE(1) + SEQ(1) + LEN(2 LE) + PAYLOAD(N) + CRC8(1)
 */

import { PROTOCOL_VERSION, MAGIC, HEADER_SIZE } from './constants.js'
import { crc8 } from './crc8.js'
import type { GraftPacket } from './types.js'

/**
 * Encode a GRAFT packet into a raw byte buffer.
 * @returns Complete packet bytes (MAGIC through CRC)
 */
export function wireEncode(packet: GraftPacket): Uint8Array {
  const payloadLen = packet.payload.length
  const total = HEADER_SIZE + payloadLen
  const buf = new Uint8Array(total)

  // MAGIC
  buf[0] = MAGIC[0]
  buf[1] = MAGIC[1]

  // VER
  buf[2] = packet.version

  // TYPE
  buf[3] = packet.type

  // SEQ
  buf[4] = packet.seq

  // LEN (little-endian uint16)
  buf[5] = payloadLen & 0xFF
  buf[6] = (payloadLen >> 8) & 0xFF

  // PAYLOAD
  buf.set(packet.payload, 7)

  // CRC8 over [VER..end of PAYLOAD] = bytes [2..7+payloadLen-1]
  const crcRegion = buf.subarray(2, 7 + payloadLen)
  buf[7 + payloadLen] = crc8(crcRegion)

  return buf
}

/**
 * Decode a raw byte buffer into a GRAFT packet.
 * @returns Decoded packet, or throws on error
 */
export function wireDecode(buf: Uint8Array): GraftPacket {
  if (buf.length < HEADER_SIZE) {
    throw new Error(`Incomplete packet: need at least ${HEADER_SIZE} bytes, got ${buf.length}`)
  }

  // Check magic
  if (buf[0] !== MAGIC[0] || buf[1] !== MAGIC[1]) {
    throw new Error(`Invalid magic: expected 0x${MAGIC[0].toString(16)}${MAGIC[1].toString(16)}, got 0x${buf[0].toString(16)}${buf[1].toString(16)}`)
  }

  const version = buf[2]
  if (version !== PROTOCOL_VERSION) {
    throw new Error(`Version mismatch: expected ${PROTOCOL_VERSION}, got ${version}`)
  }

  const type = buf[3]
  const seq = buf[4]
  const payloadLen = buf[5] | (buf[6] << 8)

  const total = HEADER_SIZE + payloadLen
  if (buf.length < total) {
    throw new Error(`Incomplete packet: need ${total} bytes, got ${buf.length}`)
  }

  // Verify CRC
  const crcRegion = buf.subarray(2, 7 + payloadLen)
  const expectedCrc = crc8(crcRegion)
  const actualCrc = buf[7 + payloadLen]

  if (expectedCrc !== actualCrc) {
    throw new Error(`CRC mismatch: expected 0x${expectedCrc.toString(16)}, got 0x${actualCrc.toString(16)}`)
  }

  const payload = buf.slice(7, 7 + payloadLen)

  return { version, type, seq, payload }
}

/**
 * Build a GraftPacket ready for encoding.
 */
export function makePacket(type: number, seq: number, payload: Uint8Array = new Uint8Array(0)): GraftPacket {
  return {
    version: PROTOCOL_VERSION,
    type,
    seq,
    payload,
  }
}

/**
 * Calculate total wire size for a given payload length.
 */
export function wirePacketSize(payloadLen: number): number {
  return HEADER_SIZE + payloadLen
}

/**
 * Try to find a complete GRAFT packet in a byte stream.
 * Scans for MAGIC bytes and attempts to decode.
 * @returns [packet, bytesConsumed] or null if no complete packet found
 */
export function wireFindPacket(buf: Uint8Array): [GraftPacket, number] | null {
  for (let i = 0; i <= buf.length - HEADER_SIZE; i++) {
    if (buf[i] !== MAGIC[0] || buf[i + 1] !== MAGIC[1]) continue

    const payloadLen = buf[i + 5] | (buf[i + 6] << 8)
    const total = HEADER_SIZE + payloadLen

    if (i + total > buf.length) return null // incomplete, need more data

    try {
      const packet = wireDecode(buf.subarray(i, i + total))
      return [packet, i + total]
    } catch {
      // CRC or version mismatch at this offset, keep scanning
      continue
    }
  }

  return null
}

File: sdk/python/src/graft/core/wire.py
"""GRAFT Wire Format — Packet encode/decode"""

from __future__ import annotations

import struct
from dataclasses import dataclass

from .constants import PROTOCOL_VERSION, MAGIC, HEADER_SIZE
from .crc8 import crc8


@dataclass
class GraftPacket:
    version: int
    type: int
    seq: int
    payload: bytes


def wire_encode(packet: GraftPacket) -> bytes:
    """Encode a GraftPacket into raw wire bytes."""
    payload_len = len(packet.payload)
    buf = bytearray(HEADER_SIZE + payload_len)

    # MAGIC
    buf[0] = MAGIC[0]
    buf[1] = MAGIC[1]
    # VER
    buf[2] = packet.version
    # TYPE
    buf[3] = packet.type
    # SEQ
    buf[4] = packet.seq
    # LEN (little-endian uint16)
    struct.pack_into("<H", buf, 5, payload_len)
    # PAYLOAD
    buf[7 : 7 + payload_len] = packet.payload
    # CRC8 over [VER..end of PAYLOAD]
    crc_region = buf[2 : 7 + payload_len]
    buf[7 + payload_len] = crc8(crc_region)

    return bytes(buf)


def wire_decode(data: bytes | bytearray) -> GraftPacket:
    """Decode raw wire bytes into a GraftPacket. Raises ValueError on error."""
    if len(data) < HEADER_SIZE:
        raise ValueError(f"Incomplete packet: need {HEADER_SIZE} bytes, got {len(data)}")

    if data[0] != MAGIC[0] or data[1] != MAGIC[1]:
        raise ValueError(f"Invalid magic: 0x{data[0]:02x}{data[1]:02x}")

    version = data[2]
    if version != PROTOCOL_VERSION:
        raise ValueError(f"Version mismatch: expected {PROTOCOL_VERSION}, got {version}")

    pkt_type = data[3]
    seq = data[4]
    payload_len = struct.unpack_from("<H", data, 5)[0]

    total = HEADER_SIZE + payload_len
    if len(data) < total:
        raise ValueError(f"Incomplete packet: need {total} bytes, got {len(data)}")

    crc_region = data[2 : 7 + payload_len]
    expected_crc = crc8(crc_region)
    actual_crc = data[7 + payload_len]

    if expected_crc != actual_crc:
        raise ValueError(f"CRC mismatch: expected 0x{expected_crc:02x}, got 0x{actual_crc:02x}")

    payload = bytes(data[7 : 7 + payload_len])
    return GraftPacket(version=version, type=pkt_type, seq=seq, payload=payload)


def make_packet(pkt_type: int, seq: int, payload: bytes = b"") -> GraftPacket:
    """Create a GraftPacket ready for encoding."""
    return GraftPacket(version=PROTOCOL_VERSION, type=pkt_type, seq=seq, payload=payload)
